data_schemes.py

```diff
@@ -0,0 +1,70 @@
+from marshmallow import Schema, fields, ValidationError
+
+
+class BoolLowerCase(fields.Field):
+    """
+    Field that serializes bool value to lower case string
+    True -> "true", False -> "false"
+    """
+    def _serialize(self, value, attr, obj, **kwargs):
+        return str(value).lower()
+
+
+class ConfirmationSchema(fields.Field):
+    """
+    If confirmations == 0, this mean that transactions are considered confirmed when "irreversible" blockchains
+    time has come.
+    If confirmations > 0, this mean that gateway need to counting "value" of next blocks after goal transaction block
+    """
+    def _serialize(self, value, attr, obj, **kwargs):
+        if value == 0:
+            return {"type": "irreversible"}
+        else:
+            return {"type": "block", "value": value}
+
+
+class AssetDataSchema(Schema):
+    class Meta:
+        ordered = True
+
+    name = fields.Str(required=True)
+    unified_cryptoasset_id = fields.Str()
+    can_deposit = BoolLowerCase()
+    can_withdraw = BoolLowerCase()
+    min_withdraw = fields.Str()
+    max_withdraw = fields.Str()
+    maker_fee = fields.Str()
+    taker_fee = fields.Str()
+
+
+class CoinDataSchema(Schema):
+    class Meta:
+        ordered = True
+
+    name = fields.Str(required=True)
+    description = fields.Str(required=True)
+    backing_coin = fields.Str()
+    symbol = fields.Str()
+
+    deposit_allowed = fields.Bool()
+    withdrawal_allowed = fields.Bool()
+
+    memo_support = fields.Bool()
+    memo_version = fields.Int(load_only=True)
+
+    precision = fields.Int()
+
+    issuer = fields.Str()
+    issuer_id = fields.Str()
+
+    gateway_wallet = fields.Str(load_only=True)
+    wallet_type = fields.Str()
+
+    min_amount = fields.Int()
+    withdraw_fee = fields.Int()
+    deposit_fee = fields.Int()
+
+    # gate_fee do not store in database, and calculated as withdraw_fee / 10 ** precision
+    gate_fee = fields.Str()
+
+    confirmations = ConfirmationSchema()
```

Comment:
Use marshmallow_dataclass for better mypy support in the future.. Example:

```python3
# data_transfer_classes.py
from typing import Type, ClassVar, List, Optional
from decimal import Decimal
import marshmallow
from marshmallow import Schema
from marshmallow_dataclass import dataclass, NewType


def AmountField(*args, **kwargs):
    return marshmallow.fields.Decimal(*args, **kwargs, as_string=True)


Amount = NewType("Amount", Decimal, field=AmountField)


@dataclass
class DataTransferClass:
    Schema: ClassVar[Type[Schema]] = Schema


# example
class HelloServer(DataTransferClass):
    first_name: str
    last_name: Optional[str] = None


# example
class HelloUser(DataTransferClass):
    motd: str


# decorators.py
from typing import Type, Union, Callable, Awaitable, Mapping, Optional
from aiohttp.web import Request, Response, HTTPException, json_response
import functools
import re
import json
from json.decoder import JSONDecodeError
from marshmallow import Schema
from marshmallow.exceptions import ValidationError as SchemaValidationError
import logging


from data_transfer_classes import DataTransferClass


# TODO: variadic callable, blocked by
# https://github.com/python/typing/issues/239
ErrorsMapHandler = Callable[[Request], Awaitable[Response]]


ErrorsMapWrapper = Callable[[ErrorsMapHandler], OriginalHandler]


def errors_map(errors: Mapping[Type[Exception],
                               Type[HTTPException]]) -> ErrorsMapWrapper:
    def to_snake_case(name: str) -> str:
        s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)

        return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()


    def handler_wrapper(handler: ErrorsMapHandler) -> OriginalHandler:
        @functools.wraps(handler)
        async def map(request: Request, *args) -> Response:
            try:
                return await handler(request, *args)
            except Exception as exception:
                for error_from, error_to in errors.items():
                    if isinstance(exception, error_from):
                        logging.getLogger('aiohttp.web').exception(exception)

                        raise error_to(body=json.dumps({
                            'error': {
                                'type': to_snake_case(error_from.__name__),
                                'description': str(exception)
                            }
                        }), content_type='application/json')

                raise exception
        return map
    return handler_wrapper

# TODO: variadic callable, blocked by
# https://github.com/python/typing/issues/239
DTOValidateHandler = Union[Callable[[Request, DataTransferClass],
                                     Awaitable[DataTransferClass]],
                           Callable[[Request], Awaitable[DataTransferClass]]]


DTOValidateWrapper = Callable[[DTOValidateHandler], OriginalHandler]


def dto_validate(dtc: Optional[DataTransferClass] = None) -> DTOValidateWrapper:
    def handler_wrapper(handler: DTOValidateHandler) -> OriginalHandler:
        @functools.wraps(handler)
        async def validate(request: Request, *args) -> Response:
            if dtc is None:
                dto = await handler(request, *args)
            else:
                try:
                    payload = await request.json()
                except JSONDecodeError as exception:
                    raise InvalidJSON(
                        f'Invalid JSON payload: {exception}'
                    )
                try:
                    dto = dtc.Schema().load(payload)
                except SchemaValidationError as exception:
                    raise InvalidJSONType(
                        f'Invalid JSON payload type: {exception}'
                    )

                dto = await handler(request, dto, *args)

            schema = dto.Schema()
            payload = schema.dump(dto)

            try:
                schema.validate(payload)
            except SchemaValidationError as exception:
                raise InternalInvalidJSONType(
                    f'Invalid JSON payload type: {exception}'
                )

            return json_response(payload)
        return validate
    return handler_wrapper

#views.py
#example
from aiohttp.web import Request, HTTPBadRequest
import logging


from data_transfer_classes import HelloServer, HelloUser
from decorators import InvalidJSON, InvalidJSONType, errors_map, dto_validate


routes = RouteTableDef()


@routes.post('/hello')
@errors_map({InvalidJSON: HTTPBadRequest, InvalidJSONType: HTTPBadRequest})
@dto_validate(HelloServer)
async def hello(request: Request, dto: HelloServer) -> HelloUser:
    logger = logging.getLogger('aiohttp.web')

    logger.info(dto)

    return HelloUser(motd="Hello, World")
```

Reply:
What is OriginalHandler class? How to declarate it?